src/frontend/Typechecker.ml

```diff
@@ -1357,19 +1327,29 @@ let rec check_if_then_else loc cf tenv pred_e s_true s_false_opt =
   let srt2 =
     ts_false_opt
     |> Option.map ~f:(fun s -> s.smeta.return_type)
-    |> Option.value ~default:NoReturnType in
-  let return_type = try_compute_ifthenelse_statement_returntype loc srt1 srt2 in
+    |> Option.value ~default:Incomplete in
+  let return_type = try_compute_ifthenelse_statement_returntype srt1 srt2 in
   mk_typed_statement ~stmt ~return_type ~loc
 
 and check_while loc cf tenv cond_e loop_body =
+  let hardcoded_true e =
+    (* heuristic for "will this loop forever" *)
+    match e.expr with
+    | Ast.IntNumeral s -> String.exists s ~f:(fun c -> c > '0' && c <= '9')
+    | _ -> false in
   let _, ts =
-    check_statement {cf with loop_depth= cf.loop_depth + 1} tenv loop_body
-  and te =
+    check_statement {cf with loop_depth= cf.loop_depth + 1} tenv loop_body in
+  let te =
     check_expression_of_int_or_real_type cf tenv cond_e
       "Condition in while-loop" in
-  mk_typed_statement
-    ~stmt:(While (te, ts))
-    ~return_type:ts.smeta.return_type ~loc
+  let return_type =
+    match (hardcoded_true te, ts.smeta.return_type) with
+    | true, Incomplete ->
```

Comment:
I think I would prefer
```ocaml
    match ts.smeta.return_type with
    | Incomplete when hardcoded_true te ->
```
or `unconditional_true` (as I mentioned on the other PR).

Reply:
That's better, yes.

I prefer `hardcoded_true` to `unconditional_true` since to me it captures more correctly the idea that we're not checking anything clever, e.g. using the partial evaluator or something like that. `(10 %/% 2 - 1)` is also unconditional, but it is not hard coded (IMO)